File: vivordo_health/functions/heart_health_score.js
"use strict";

const HEART_HEALTH_BASELINE_WINDOW_DAYS = 28;
const HEART_HEALTH_MINIMUM_BASELINE_DAYS = 7;

const valid = (value) => Number.isFinite(value) && value > 0;

const median = (values) => {
  const sorted = [...values].sort((a, b) => a - b);
  const middle = Math.floor(sorted.length / 2);
  return sorted.length % 2 === 1 ?
    sorted[middle] :
    (sorted[middle - 1] + sorted[middle]) / 2;
};

const personalizedScore = ({
  current,
  history,
  lowerIsBetter,
  minimumScale,
  pointsPerDeviation,
}) => {
  const values = history.filter(valid);
  if (!valid(current) || values.length < HEART_HEALTH_MINIMUM_BASELINE_DAYS) {
    return null;
  }
  const baseline = median(values);
  const deviation = median(values.map((value) => Math.abs(value - baseline)));
  const scale = Math.max(deviation * 1.4826, minimumScale);
  const standardizedDifference = (current - baseline) / scale;
  const direction = lowerIsBetter ? -1 : 1;
  return Math.max(0, Math.min(
      100,
      80 + standardizedDifference * direction * pointsPerDeviation,
  ));
};

/**
 * Calculates personalized Heart Health while redistributing missing weights.
 *
 * @param {Object} current Current resting HR, SDNN HRV, and quiet HR.
 * @param {Object[]} history Up to 28 earlier daily signal objects.
 * @return {Object} Score, component scores, baseline state, and confidence.
 */
function calculateHeartHealthScore(current, history) {
  const restingHistory = history.map((day) => day.restingHeartRate);
  const hrvHistory = history.map((day) => day.hrvSdnn);
  const quietHistory = history.map((day) => day.quietHeartRate);
  const restingHeartRateScore = personalizedScore({
    current: current.restingHeartRate,
    history: restingHistory,
    lowerIsBetter: true,
    minimumScale: 3,
    pointsPerDeviation: 12,
  });
  const hrvScore = personalizedScore({
    current: current.hrvSdnn,
    history: hrvHistory,
    lowerIsBetter: false,
    minimumScale: 5,
    pointsPerDeviation: 10,
  });
  const quietHeartRateScore = personalizedScore({
    current: current.quietHeartRate,
    history: quietHistory,
    lowerIsBetter: true,
    minimumScale: 3,
    pointsPerDeviation: 10,
  });
  const components = [
    {score: restingHeartRateScore, weight: 0.50},
    {score: hrvScore, weight: 0.30},
    {score: quietHeartRateScore, weight: 0.20},
  ];
  const scored = components.filter((component) => component.score !== null);
  const availableWeight = scored.reduce(
      (total, component) => total + component.weight,
      0,
  );
  const score = availableWeight === 0 ? null : scored.reduce(
      (total, component) => total + component.score * component.weight,
      0,
  ) / availableWeight;
  const availableSignals = [
    current.restingHeartRate,
    current.hrvSdnn,
    current.quietHeartRate,
  ].filter(valid).length;
  const baselineCounts = [
    restingHeartRateScore === null ? null : restingHistory.filter(valid).length,
    hrvScore === null ? null : hrvHistory.filter(valid).length,
    quietHeartRateScore === null ? null : quietHistory.filter(valid).length,
  ].filter((value) => value !== null);
  const allCounts = [restingHistory, hrvHistory, quietHistory]
      .map((values) => values.filter(valid).length);
  const baselineDays = baselineCounts.length > 0 ?
    Math.min(...baselineCounts) : Math.max(...allCounts);
  const scoredSignals = scored.length;
  const confidence = scoredSignals === 3 && baselineDays >= 14 ?
    "high" : scoredSignals >= 2 || (scoredSignals === 1 && baselineDays >= 14) ?
    "medium" : "low";

  return {
    score,
    restingHeartRateScore,
    hrvScore,
    quietHeartRateScore,
    availableSignals,
    scoredSignals,
    baselineDays,
    confidence,
    isBuildingBaseline: availableSignals > 0 && score === null,
  };
}

module.exports = {
  calculateHeartHealthScore,
  HEART_HEALTH_BASELINE_WINDOW_DAYS,
  HEART_HEALTH_MINIMUM_BASELINE_DAYS,
};

File: vivordo_health/lib/src/utils/heart_health_score.dart
import 'dart:math' as math;

const int heartHealthBaselineWindowDays = 28;
const int heartHealthMinimumBaselineDays = 7;

class HeartHealthSignals {
  const HeartHealthSignals({
    this.restingHeartRate,
    this.hrvSdnn,
    this.quietHeartRate,
  });

  final double? restingHeartRate;
  final double? hrvSdnn;
  final double? quietHeartRate;
}

enum HeartHealthConfidence { low, medium, high }

class HeartHealthScoreResult {
  const HeartHealthScoreResult({
    required this.score,
    required this.restingHeartRateScore,
    required this.hrvScore,
    required this.quietHeartRateScore,
    required this.availableSignals,
    required this.scoredSignals,
    required this.baselineDays,
    required this.confidence,
  });

  final double? score;
  final double? restingHeartRateScore;
  final double? hrvScore;
  final double? quietHeartRateScore;
  final int availableSignals;
  final int scoredSignals;
  final int baselineDays;
  final HeartHealthConfidence confidence;

  bool get isBuildingBaseline => availableSignals > 0 && score == null;
}

/// Produces a personalized Heart Health score from cardiovascular trends.
///
/// A user's own recent median is the neutral reference point (80/100). Resting
/// and quiet heart rates score better when modestly below their usual level;
/// HRV scores better when above its usual level. Median absolute deviation
/// makes the comparison robust to isolated spikes. Signals with fewer than
/// seven prior days are excluded and the remaining weights are normalized.
HeartHealthScoreResult calculateHeartHealthScore({
  required HeartHealthSignals current,
  required List<HeartHealthSignals> history,
}) {
  final restingHistory = _values(history.map((day) => day.restingHeartRate));
  final hrvHistory = _values(history.map((day) => day.hrvSdnn));
  final quietHistory = _values(history.map((day) => day.quietHeartRate));

  final restingScore = _personalizedScore(
    current: current.restingHeartRate,
    history: restingHistory,
    lowerIsBetter: true,
    minimumScale: 3,
    pointsPerDeviation: 12,
  );
  final hrvScore = _personalizedScore(
    current: current.hrvSdnn,
    history: hrvHistory,
    lowerIsBetter: false,
    minimumScale: 5,
    pointsPerDeviation: 10,
  );
  final quietScore = _personalizedScore(
    current: current.quietHeartRate,
    history: quietHistory,
    lowerIsBetter: true,
    minimumScale: 3,
    pointsPerDeviation: 10,
  );

  const weights = [0.50, 0.30, 0.20];
  final scores = [restingScore, hrvScore, quietScore];
  var weightedScore = 0.0;
  var availableWeight = 0.0;
  for (var index = 0; index < scores.length; index++) {
    final score = scores[index];
    if (score == null) continue;
    weightedScore += score * weights[index];
    availableWeight += weights[index];
  }

  final availableSignals = [
    current.restingHeartRate,
    current.hrvSdnn,
    current.quietHeartRate,
  ].where(_valid).length;
  final scoredSignals = scores.whereType<double>().length;
  final baselineCounts = <int>[
    if (restingScore != null) restingHistory.length,
    if (hrvScore != null) hrvHistory.length,
    if (quietScore != null) quietHistory.length,
  ];
  final baselineDays = baselineCounts.isEmpty
      ? math.max(
          restingHistory.length,
          math.max(hrvHistory.length, quietHistory.length),
        )
      : baselineCounts.reduce(math.min);
  final confidence = scoredSignals == 3 && baselineDays >= 14
      ? HeartHealthConfidence.high
      : scoredSignals >= 2 || (scoredSignals == 1 && baselineDays >= 14)
      ? HeartHealthConfidence.medium
      : HeartHealthConfidence.low;

  return HeartHealthScoreResult(
    score: availableWeight == 0
        ? null
        : (weightedScore / availableWeight).clamp(0.0, 100.0),
    restingHeartRateScore: restingScore,
    hrvScore: hrvScore,
    quietHeartRateScore: quietScore,
    availableSignals: availableSignals,
    scoredSignals: scoredSignals,
    baselineDays: baselineDays,
    confidence: confidence,
  );
}

List<double> _values(Iterable<double?> values) => values
    .where(_valid)
    .map((value) => value!.toDouble())
    .toList(growable: false);

bool _valid(double? value) => value != null && value.isFinite && value > 0;

double? _personalizedScore({
  required double? current,
  required List<double> history,
  required bool lowerIsBetter,
  required double minimumScale,
  required double pointsPerDeviation,
}) {
  if (!_valid(current) || history.length < heartHealthMinimumBaselineDays) {
    return null;
  }
  final baseline = _median(history);
  final deviations = history
      .map((value) => (value - baseline).abs())
      .toList(growable: false);
  final scale = math.max(_median(deviations) * 1.4826, minimumScale);
  final standardizedDifference = (current! - baseline) / scale;
  final direction = lowerIsBetter ? -1.0 : 1.0;
  return (80 + standardizedDifference * direction * pointsPerDeviation).clamp(
    0.0,
    100.0,
  );
}

double _median(List<double> values) {
  final sorted = [...values]..sort();
  final middle = sorted.length ~/ 2;
  if (sorted.length.isOdd) return sorted[middle];
  return (sorted[middle - 1] + sorted[middle]) / 2;
}
